### PR TITLE
Query tools

### PR DESCRIPTION
**Pull request by isaac-looker**
See the demo: https://github.com/Spelman-College/spelman-dashboard/pull/60

- adding query logic
  - creates an index(an array) of dcids using inferred dimensions and inferred categories
  - categories and dimensions must be known ahead of time
  - for a given query (1 or more categories, and 1 or more dimenions per category) create the cartesian product of category/dimension scan the dcid index to acquire each relevant dcid string.
- add Jest testing framework

### CODE REVIEW

**Review by cernst122 — APPROVED** [review]
Thanks for the thorough tests. LGTM for a first iteration. In order to avoid ballooning complexity, I'd endorse playing hard and fast with hardcoded values for our known sets wherever doing so will be simpler. 

I'm fine with whatever you prefer re: general or hardcoded approach as long as we follow the below principles:
- We should not only support all permutations of properties, but also the absence of properties (iff the data set has a total field that allows us to do this). 
- When a multiple values for the same mutually exclusive property are selected, this should trigger separate API calls (e.g. pet="bird", tool="hammer", "ruler" could be two line graphs, one for "bird hammer" and one for "bird ruler").

**Comment by isaac-looker** [comment]
> Thanks for the thorough tests. LGTM for a first iteration. In order to avoid ballooning complexity, I'd endorse playing hard and fast with hardcoded values for our known sets wherever doing so will be simpler.
> 
> I'm fine with whatever you prefer re: general or hardcoded approach as long as we follow the below principles:
> 
> * We should not only support all permutations of properties, but also the absence of properties (iff the data set has a total field that allows us to do this).

I added an example to the [demo](https://github.com/Spelman-College/spelman-dashboard/pull/60/commits/88bea6bf8e9cf4dfaf32ca9ac73823726b8317d7)- output shown here

```
npx tsc example.ts --target esnext  --module commonjs && node example.js
gender
ethnicity
occupation
All genders, Black and Asian, SOCLifeScientistsOccupation, SOCMathematicalScienceOccupation
[ 
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Male_SOCLifeScientistsOccupation_Black',
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Male_SOCMathematicalScienceOccupation_Black',
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Male_SOCLifeScientistsOccupation_Asian',
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Male_SOCMathematicalScienceOccupation_Asian',
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Female_SOCLifeScientistsOccupation_Black',
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Female_SOCMathematicalScienceOccupation_Black',
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Female_SOCLifeScientistsOccupation_Asian',
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Female_SOCMathematicalScienceOccupation_Asian'
]
Only Female
[ 
  'Count_Person_ScienceAndEngineeringRelatedMajor_EducationalAttainmentDoctorateDegree_Female'
]
```

Does this demonstrate your requirement?

> * When a multiple values for the same mutually exclusive property are selected, this should trigger separate API calls (e.g. pet="bird", tool="hammer", "ruler" could be two line graphs, one for "bird hammer" and one for "bird ruler").

This PR adds the ability to acquire the union of queries across mutually exclusive categories, using a single `QuerySet`. I believe what you're referring to is the ability to make more than one query, each containing the same category; in this case, we would 'compile' 2 separate queries(and `QuerySet`s) and render each. Does this seem like an approach that would work?

**Comment by cernst122** [comment]
Demo output LGTM for that requirement. 

Compiling multiple times is what I was thinking, thanks for confirming we're on the same page.

File: frontend/src/queries/dcids.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            }
+
+            const cat = dim2cat[dim]
+            if (cat === undefined) {
+                throw Error(`missing dimension: ${dim} in the dimension to category map`)
+            }
+            unUsedCats.delete(cat)
+            this.dimensions.add(`${cat}:${dim}`)
+        })
+        // Add the query all for any missing categories in this dcid
```
[the remark this answers is not in the document]
Comment (isaac-looker) [reply frontend/src/queries/dcids.ts]: I've updated the comment and https://github.com/Spelman-College/spelman-dashboard/pull/60 (the example)- output shown [here](https://gist.github.com/isaac-looker/0fc2d0a0e8035f6968f857fde0a85bc0)

**Pull request merged**